oml/datasets/images.py

```diff
@@ -0,0 +1,428 @@
+        img_bytes = self.read_bytes(self._paths[idx])
+        img = self._f_imread(img_bytes)
+
+        im_h, im_w = img.shape[:2] if isinstance(img, np.ndarray) else img.size[::-1]
+
+        if (self._bboxes is not None) and (self._bboxes[idx] is not None):
+            x1, y1, x2, y2 = self._bboxes[idx]
+        else:
+            x1, y1, x2, y2 = 0, 0, im_w, im_h
+
+        if isinstance(self._transform, albu.Compose):
+            img = img[y1:y2, x1:x2, :]
+            image_tensor = self._transform(image=img)["image"]
+        else:
+            # torchvision.transforms
+            img = img.crop((x1, y1, x2, y2))
+            image_tensor = self._transform(img)
+
+        item = {
+            self.input_tensors_key: image_tensor,
+            self.index_key: idx,
+        }
+
+        if self.extra_data:
+            for key, record in self.extra_data.items():
+                if key in item:
+                    raise ValueError(f"<extra_data> and dataset share the same key: {key}")
+                else:
+                    item[key] = record[idx]
+
+        # todo 522: remove
+        item[self.x1_key] = x1
+        item[self.y1_key] = y1
+        item[self.x2_key] = x2
+        item[self.y2_key] = y2
+        item[self.paths_key] = self._paths[idx]
+
+        return item
+
+    def __len__(self) -> int:
+        return len(self._paths)
+
+    def visualize(self, idx: int, color: TColor = BLACK) -> np.ndarray:
+        bbox = torch.tensor(self._bboxes[idx]) if (self._bboxes is not None) else torch.tensor([torch.nan] * 4)
+        image = get_img_with_bbox(im_path=self._paths[idx], bbox=bbox, color=color)
+        image = square_pad(image)
+
+        return image
+
+    # todo 522: remove
+    @property
+    def bboxes_keys(self) -> Tuple[str, ...]:
+        return self.x1_key, self.y1_key, self.x2_key, self.y2_key
+
+
+class ImageLabeledDataset(ImageBaseDataset, ILabeledDataset):
+    """
+    The dataset of images having their ground truth labels.
+
+    """
+
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        extra_data: Optional[Dict[str, Any]] = None,
+        dataset_root: Optional[Union[str, Path]] = None,
+        transform: Optional[albu.Compose] = None,
+        f_imread: Optional[TImReader] = None,
+        cache_size: Optional[int] = 0,
+        input_tensors_key: str = INPUT_TENSORS_KEY,
+        labels_key: str = LABELS_KEY,
+        index_key: str = INDEX_KEY,
+        # todo 522: remove
+        paths_key: str = PATHS_KEY,
+        categories_key: Optional[str] = CATEGORIES_KEY,
+        sequence_key: Optional[str] = SEQUENCE_KEY,
+        x1_key: str = X1_KEY,
+        x2_key: str = X2_KEY,
+        y1_key: str = Y1_KEY,
+        y2_key: str = Y2_KEY,
+    ):
+        assert (LABELS_COLUMN in df) and (PATHS_COLUMN in df), "There are only 2 required columns."
+        self.labels_key = labels_key
+        self.df = df
+
+        extra_data = {} if extra_data is None else extra_data
```

Comment:
thank you for finding this inconsistency. I've decided to keep {} eveywhere because than we can call `dataset.extra_data.ket(k, None)` without extra ifs